src/toil_lib/programs.py

```diff
@@ -124,3 +125,282 @@ def _fix_permissions(base_docker_call, tool, work_dir):
     stat = os.stat(work_dir)
     command = base_docker_call + [tool] + ['-R', '{}:{}'.format(stat.st_uid, stat.st_gid), '/data']
     subprocess.check_call(command)
+
+
+class Docker(object):
+    """
+    A builder (https://en.wikipedia.org/wiki/Builder_pattern) for Popen objects representing a
+    Docker client invocation. When building a Popen object, it distinguishes between three
+    layers, the `popen` layer, the `docker` layer and the `extra` layer. The `popen` layer
```

Comment:
I don't like `extra` as a layer name, because it is never referred to as such in Docker docs.  Leaving that off is perhaps not clear enough?
```python
(Docker()
  .docker_command('run') # invoke `docker run`
  .docker_params('ubuntu', interactive=True) # pass `--interactive` and run `ubuntu` image
  .docker_params(rm=True)  # docker() is cumulative, also pass `--rm`
  .command('foo') # invoke `foo` command inside container
  .params('/', bar=42) # pass `--bar=42 /` to foo
  .popen_params(cwd='/foo/bar') # pass `cwd='/foo/bar'` to the Popen constructor
  .work_dir('/foo/bar')  # same
  .build_popen_call())
```